internal/identity/agent.go

```diff
@@ -0,0 +1,41 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package identity // import "gopkg.in/juju/charmstore.v5-unstable/internal/identity"
+
+import (
+	"net/url"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/macaroon-bakery.v1/bakery"
+)
+
+type loginMethods struct {
+	Agent string `json:"agent"`
+}
+
+type agentLoginRequest struct {
+	Username  string            `json:"username"`
+	PublicKey *bakery.PublicKey `json:"public_key"`
+}
+
+func (c *Client) agentLogin(u *url.URL) error {
```

Comment:
// agentLogin implements a non-interactive visitURL function.

?

I wonder if this should check whether the url is rooted at c.URL
so that we don't try to use the identity username on other third
party services. Alternatively we could have some kind of map
from URLs to usernames (not really that useful for us given that
we've only got one third party service we care about here)

I still think that setting a cookie at the idm location containing
the declared username and password would have been a reasonable
solution (and quite a bit less code).